### PR TITLE
NodeNameResolver: Throw exception in BC breaking path

### PR DESCRIPTION
**Pull request by staabm**

### CODE REVIEW

File: packages/NodeNameResolver/NodeNameResolver.php [hunk 1]
```diff
@@ -181,6 +186,12 @@ public function isStringName(string $resolvedName, string $desiredName): bool
             return $desiredName === $resolvedName;
         }
 
+        foreach(self::REGEX_WILDCARD_CHARS as $char) {
+            if (str_contains($desiredName, $char)) {
+                throw new ShouldNotHappenException('Matching of regular expressions is no longer supported');
```
Comment (TomasVotruba) [inline packages/NodeNameResolver/NodeNameResolver.php]: Thanks :+1: 

I'd add most common replacement solution like:
 "Use exact compare of $this->getName() and str_ends_with() or str_starts_with() instead"

Comment (staabm) [reply packages/NodeNameResolver/NodeNameResolver.php]: good point, fixed

**Comment by staabm** [comment]
the failling build tells me we have still some work left in the codebase to compensate the removal of the regex matching ;)

**Comment by samsonasik** [comment]
@staabm the target code can be updated with:

```php
if (str_contains($name, "*")) {
    fnmatch()...

else
   $this->isName()

**Comment by samsonasik** [comment]
@staabm please rebase, the fnmatch handling should already covered at PR:

- https://github.com/rectorphp/rector-src/pull/4988
- https://github.com/rectorphp/rector-src/pull/4990
- https://github.com/rectorphp/rector-src/pull/4991

**Comment by staabm** [comment]
rebased.

**Comment by samsonasik** [comment]
@staabm please rebase once more :)

**Review by samsonasik — APPROVED** [review]
It seems all green now :)

File: packages/NodeNameResolver/NodeNameResolver.php [hunk 2]
```diff
@@ -181,6 +186,12 @@ public function isStringName(string $resolvedName, string $desiredName): bool
             return $desiredName === $resolvedName;
         }
 
+        foreach(self::REGEX_WILDCARD_CHARS as $char) {
+            if (str_contains($desiredName, $char)) {
+                throw new ShouldNotHappenException('Matching of regular expressions is no longer supported. Use $this->getName() and compare with e.g. str_ends_with() or str_starts_with() instead.');
```
Comment (samsonasik) [inline packages/NodeNameResolver/NodeNameResolver.php:191]: fnmatch should mentioned:

```suggestion
                throw new ShouldNotHappenException('Matching of regular expressions is no longer supported. Use $this->getName() and compare with e.g. str_ends_with(), str_starts_with(), or fnmatch() instead.');
```

Comment (samsonasik) [reply packages/NodeNameResolver/NodeNameResolver.php:191]: fnmatch example https://3v4l.org/MkqgX

Comment (samsonasik) [reply packages/NodeNameResolver/NodeNameResolver.php:191]: hm.., this must be moved in last resort after `return strcasecmp($resolvedName, $desiredName) === 0`

Comment (TomasVotruba) [reply packages/NodeNameResolver/NodeNameResolver.php:191]: I want to exclude fnmatch() on purpose, as it's mostly related to paths, not node names. Also it's pain to use as * works only with specific constant :)

LGTM :+1:

File: packages/NodeNameResolver/NodeNameResolver.php [hunk 3]
```diff
@@ -181,6 +186,12 @@ public function isStringName(string $resolvedName, string $desiredName): bool
             return $desiredName === $resolvedName;
         }
 
+        foreach(self::REGEX_WILDCARD_CHARS as $char) {
```
Comment (samsonasik) [inline packages/NodeNameResolver/NodeNameResolver.php:189]: move this after `strcasecmp($resolvedName, $desiredName) === 0`

```php
       if (strcasecmp($resolvedName, $desiredName) === 0) {
             return true;
       }
      
       // add here...
      foreach(self::REGEX_WILDCARD_CHARS as $char) {
      }
      
       return false;
```

**Comment by TomasVotruba** [comment]
Thank you :clap:

File: packages/NodeNameResolver/NodeNameResolver.php (hunk 3 above)
Comment (samsonasik) [reply packages/NodeNameResolver/NodeNameResolver.php:189]: I will create new PR to move this after last resort to avoid unnecessary loop when possible.

**Pull request merged**